# EXERCISE TASK 1

## SETUP

In [1]:
!pip install -q datasets

In [2]:
from datasets import load_dataset,get_dataset_infos, Dataset

## Part 1: loading a dataset from the HF hub

Using the `load_dataset` function of the `datasets` library, load each of the following datasets in turn`

*    stanfordnlp/imdb
*    eriktks/conll2003
*    openai/gsm8k

ANSWER QUESTIONS:

* What NLP task is the dataset intended for (e.g. syntactic analysis, toxicity detection, etc.)? (You may need to refer to the documentation of the dataset for this.)
* What parts is the dataset split into (e.g. train, test) and how many examples does each contain?
* What features (e.g. text, label) does the dataset have? (Try to understand how these relate to the NLP task the dataset is intended for.)
* What is the first item in the training set of the dataset?


### stanfordnlp/imdb

In [3]:
imdb_string = 'stanfordnlp/imdb'
imdb = load_dataset(imdb_string)

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


What NLP task is the dataset intended for (e.g. syntactic analysis, toxicity detection, etc.)? (You may need to refer to the documentation of the dataset for this.)

In [4]:
get_dataset_infos(imdb_string)

{'plain_text': DatasetInfo(description='', citation='', homepage='', license='', features={'text': Value(dtype='string', id=None), 'label': ClassLabel(names=['neg', 'pos'], id=None)}, post_processed=None, supervised_keys=None, builder_name='parquet', dataset_name='imdb', config_name='plain_text', version=0.0.0, splits={'train': SplitInfo(name='train', num_bytes=33435948, num_examples=25000, shard_lengths=None, dataset_name='imdb'), 'test': SplitInfo(name='test', num_bytes=32653810, num_examples=25000, shard_lengths=None, dataset_name='imdb'), 'unsupervised': SplitInfo(name='unsupervised', num_bytes=67113044, num_examples=50000, shard_lengths=None, dataset_name='imdb')}, download_checksums={'hf://datasets/stanfordnlp/imdb@e6281661ce1c48d982bc483cf8a173c1bbeb5d31/plain_text/train-00000-of-00001.parquet': {'num_bytes': 20979968, 'checksum': None}, 'hf://datasets/stanfordnlp/imdb@e6281661ce1c48d982bc483cf8a173c1bbeb5d31/plain_text/test-00000-of-00001.parquet': {'num_bytes': 20470363, 'chec

Has no description, but it is (quote from hf dataset card: https://huggingface.co/datasets/stanfordnlp/imdb):

>Large Movie Review Dataset. This is a dataset for binary sentiment classification containing substantially more data than previous benchmark datasets. We provide a set of 25,000 highly polar movie reviews for training, and 25,000 for testing. There is additional unlabeled data for use as well.

What parts is the dataset split into (e.g. train, test) and how many examples does each contain?

In [5]:
imdb

DatasetDict({
    train: Dataset({
        features: ['text', 'label'],
        num_rows: 25000
    })
    test: Dataset({
        features: ['text', 'label'],
        num_rows: 25000
    })
    unsupervised: Dataset({
        features: ['text', 'label'],
        num_rows: 50000
    })
})

Dataset is split to train, test and unsupervised and holds 25000, 25000, 50000 examples respectively

What features (e.g. text, label) does the dataset have? (Try to understand how these relate to the NLP task the dataset is intended for.)

In [6]:
print(imdb['train'][0]['text'])
print(imdb['train'][0]['label'])

I rented I AM CURIOUS-YELLOW from my video store because of all the controversy that surrounded it when it was first released in 1967. I also heard that at first it was seized by U.S. customs if it ever tried to enter this country, therefore being a fan of films considered "controversial" I really had to see this for myself.<br /><br />The plot is centered around a young Swedish drama student named Lena who wants to learn everything she can about life. In particular she wants to focus her attentions to making some sort of documentary on what the average Swede thought about certain political issues such as the Vietnam War and race issues in the United States. In between asking politicians and ordinary denizens of Stockholm about their opinions on politics, she has sex with her drama teacher, classmates, and married men.<br /><br />What kills me about I AM CURIOUS-YELLOW is that 40 years ago, this was considered pornographic. Really, the sex and nudity scenes are few and far between, eve

Dataset entries have features of `text` that represents the actual review text of the item and `label` as the representation of negative = 0, or positive = 1 review of the product

What is the first item in the training set of the dataset?

In [7]:
imdb['train'][0]

{'text': 'I rented I AM CURIOUS-YELLOW from my video store because of all the controversy that surrounded it when it was first released in 1967. I also heard that at first it was seized by U.S. customs if it ever tried to enter this country, therefore being a fan of films considered "controversial" I really had to see this for myself.<br /><br />The plot is centered around a young Swedish drama student named Lena who wants to learn everything she can about life. In particular she wants to focus her attentions to making some sort of documentary on what the average Swede thought about certain political issues such as the Vietnam War and race issues in the United States. In between asking politicians and ordinary denizens of Stockholm about their opinions on politics, she has sex with her drama teacher, classmates, and married men.<br /><br />What kills me about I AM CURIOUS-YELLOW is that 40 years ago, this was considered pornographic. Really, the sex and nudity scenes are few and far be

### eriktks/conll2003

In [8]:
conll_string = 'eriktks/conll2003'
conll = load_dataset(conll_string,trust_remote_code=True)

What NLP task is the dataset intended for (e.g. syntactic analysis, toxicity detection, etc.)? (You may need to refer to the documentation of the dataset for this.)

In [9]:
conllinfo = get_dataset_infos(conll_string)
print(conllinfo['conll2003'].description)

The shared task of CoNLL-2003 concerns language-independent named entity recognition. We will concentrate on
four types of named entities: persons, locations, organizations and names of miscellaneous entities that do
not belong to the previous three groups.

The CoNLL-2003 shared task data files contain four columns separated by a single space. Each word has been put on
a separate line and there is an empty line after each sentence. The first item on each line is a word, the second
a part-of-speech (POS) tag, the third a syntactic chunk tag and the fourth the named entity tag. The chunk tags
and the named entity tags have the format I-TYPE which means that the word is inside a phrase of type TYPE. Only
if two phrases of the same type immediately follow each other, the first word of the second phrase will have tag
B-TYPE to show that it starts a new phrase. A word with tag O is not part of a phrase. Note the dataset uses IOB2
tagging scheme, whereas the original dataset uses IOB1.

For 

What parts is the dataset split into (e.g. train, test) and how many examples does each contain?

Dataset is split to train, validation and test and holds 14041, 3250, 3250 examples respectively

In [10]:
conll

DatasetDict({
    train: Dataset({
        features: ['id', 'tokens', 'pos_tags', 'chunk_tags', 'ner_tags'],
        num_rows: 14041
    })
    validation: Dataset({
        features: ['id', 'tokens', 'pos_tags', 'chunk_tags', 'ner_tags'],
        num_rows: 3250
    })
    test: Dataset({
        features: ['id', 'tokens', 'pos_tags', 'chunk_tags', 'ner_tags'],
        num_rows: 3453
    })
})

What features (e.g. text, label) does the dataset have? (Try to understand how these relate to the NLP task the dataset is intended for.)


Dataset entries have features of id that represents the id of the example, tokens that belong to that sentence. part-of-speech (POS) tag\


What is the first item in the training set of the dataset?

In [11]:
conll['train'][0]

{'id': '0',
 'tokens': ['EU',
  'rejects',
  'German',
  'call',
  'to',
  'boycott',
  'British',
  'lamb',
  '.'],
 'pos_tags': [22, 42, 16, 21, 35, 37, 16, 21, 7],
 'chunk_tags': [11, 21, 11, 12, 21, 22, 11, 12, 0],
 'ner_tags': [3, 0, 7, 0, 0, 0, 7, 0, 0]}

### openai/gsm8k

In [12]:
gsm8k_string = 'openai/gsm8k'
gsm8k = load_dataset(gsm8k_string,"main",trust_remote_code=True)

What NLP task is the dataset intended for (e.g. syntactic analysis, toxicity detection, etc.)? (You may need to refer to the documentation of the dataset for this.)

In [13]:
gsm8k_info = get_dataset_infos(gsm8k_string)
gsm8k_info['main'].description

''

In [14]:
gsm8k

DatasetDict({
    train: Dataset({
        features: ['question', 'answer'],
        num_rows: 7473
    })
    test: Dataset({
        features: ['question', 'answer'],
        num_rows: 1319
    })
})

What parts is the dataset split into (e.g. train, test) and how many examples does each contain?

Dataset is split to train and test and holds 7473 and 1319 examples respectively

What features (e.g. text, label) does the dataset have? (Try to understand how these relate to the NLP task the dataset is intended for.)

In [15]:
print(gsm8k['train'][0]['question'])
print(gsm8k['train'][0]['answer'])

Natalia sold clips to 48 of her friends in April, and then she sold half as many clips in May. How many clips did Natalia sell altogether in April and May?
Natalia sold 48/2 = <<48/2=24>>24 clips in May.
Natalia sold 48+24 = <<48+24=72>>72 clips altogether in April and May.
#### 72


What is the first item in the training set of the dataset?

In [16]:
gsm8k['train'][0]

{'question': 'Natalia sold clips to 48 of her friends in April, and then she sold half as many clips in May. How many clips did Natalia sell altogether in April and May?',
 'answer': 'Natalia sold 48/2 = <<48/2=24>>24 clips in May.\nNatalia sold 48+24 = <<48+24=72>>72 clips altogether in April and May.\n#### 72'}

## Part 2: creating a dataset from your own data

In [22]:
!wget http://dl.turkunlp.org/TKO_8964_2023/news-fi-2021.jsonl

--2025-01-15 11:13:51--  http://dl.turkunlp.org/TKO_8964_2023/news-fi-2021.jsonl
Resolving dl.turkunlp.org (dl.turkunlp.org)... 195.148.30.23
Connecting to dl.turkunlp.org (dl.turkunlp.org)|195.148.30.23|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 36139303 (34M) [application/octet-stream]
Saving to: ‘news-fi-2021.jsonl.2’

news-fi-2021.jsonl. 100%[===================>]  34.46M  8.69MB/s    in 4.0s    

2025-01-15 11:13:55 (8.69 MB/s) - ‘news-fi-2021.jsonl.2’ saved [36139303/36139303]



In [23]:
dataset = load_dataset("json", data_files="news-fi-2021.jsonl")
ds = dataset['train'].train_test_split(test_size=0.2)

In [24]:
ds

DatasetDict({
    train: Dataset({
        features: ['summary', 'tags', 'text', 'timestamp', 'title', 'url'],
        num_rows: 5985
    })
    test: Dataset({
        features: ['summary', 'tags', 'text', 'timestamp', 'title', 'url'],
        num_rows: 1497
    })
})

In [25]:
ds['train'][0]

{'summary': 'Vaikka osalla abeista on jo kaksi koronarokotusta, tänään maanantaina alkavien ylioppilaskirjoitusten koronasäännöt pysyvät tiukkoina: osallistua saa vain oireettomana. Kahdesti rokotetulla on kuitenkin mahdollisuus välttyä karanteenilta.',
 'tags': ['Ennaltaehkäisevä terveydenhuolto',
  'Etelä-Pohjanmaa',
  'Keski-Pohjanmaa',
  'Peruskoulut ja lukiot',
  'Pohjanmaa',
  'Ressun lukio',
  'Seinäjoki',
  'Vaasa',
  'abiturientit',
  'immunisaatio',
  'injektiot',
  'karanteeni',
  'koulut',
  'lukio',
  'rehtorit',
  'rokotteet',
  'tartuntataudit',
  'ylioppilaskirjoitukset'],
 'text': 'Maanantaina 13.9. alkavat ylioppilaskirjoitukset ovat jo neljännet ”koronakirjoitukset”. Vaikka yhteiskunta alkaa vähitellen vapauttaa rajoituksia, ylioppilastutkintolautakunnan kanta on edelleen tiukka. \n – Jos on koronavirukseen liittyviä oireita, ei saa osallistua ylioppilaskirjoituksiin, sanoo ylioppilastutkintolautakunnan pääsihteeri  Taina Tähkä . Myöskään valvojat eivät voi osallistu

What NLP tasks could the dataset be used for?

Dataset could be used to train an classifier to classify news articles to different classes, dataset could be used to generate summaries of blocks of texts

What features does the dataset have?

In [28]:
ds

DatasetDict({
    train: Dataset({
        features: ['summary', 'tags', 'text', 'timestamp', 'title', 'url'],
        num_rows: 5985
    })
    test: Dataset({
        features: ['summary', 'tags', 'text', 'timestamp', 'title', 'url'],
        num_rows: 1497
    })
})

Dataset has following features:

* summary: Summary of the news article
* tags: classes that the news article has been assigned to
* text: actual news article
* timestamp: time of the news article
* title: the title of the news article
* url: the location of the news article in the internet

How many space-separated words do the texts of the dataset contain in total?

In [26]:
splits = ['train', 'test']
words = 0

for s in splits:
  for ex in ds[s]:
    ex_l_words = len(ex['text'].split(' '))
    words += ex_l_words
print(f'Count of space-separated words in the _text_ feature of the dataset entries: {words}')

Count of space-separated words in the _text_ feature of the dataset entries: 3670842
